tensorflow_quantum/core/ops/tfq_simulate_state_op.cc

```diff
@@ -131,23 +159,62 @@ class TfqSimulateStateOp : public tensorflow::OpKernel {
 
         if (start < crossover) {
           for (uint64_t j = 0; j < upper; j++) {
-            output_tensor(i, j) = ss.GetAmpl(sv, j);
+            (*output_tensor)(i, j) = ss.GetAmpl(sv, j);
           }
         }
         for (uint64_t j = upper; j < end; j++) {
-          output_tensor(i, j) = std::complex<float>(-2, 0);
+          (*output_tensor)(i, j) = std::complex<float>(-2, 0);
         }
       };
       const int num_cycles_copy = 50;
       context->device()->tensorflow_cpu_worker_threads()->workers->ParallelFor(
           uint64_t(1) << max_num_qubits, num_cycles_copy, copy_f);
     }
     sv.release();
-    programs.clear();
-    num_qubits.clear();
-    maps.clear();
-    qsim_circuits.clear();
-    fused_circuits.clear();
+  }
+
+  void ComputeSmall(
+      const std::vector<int>& num_qubits, const int max_num_qubits,
+      const std::vector<std::vector<qsim::GateFused<QsimGate>>>& fused_circuits,
+      tensorflow::OpKernelContext* context,
+      tensorflow::TTypes<std::complex<float>, 1, long int>::Matrix*
+          output_tensor) {
+    const auto tfq_for = qsim::SequentialFor(1);
+    using Simulator = qsim::Simulator<const qsim::SequentialFor&>;
+    using StateSpace = Simulator::StateSpace;
+    using State = StateSpace::State;
+
+    auto DoWork = [&](int start, int end) {
+      int largest_nq = 1;
+      State sv = StateSpace(largest_nq, tfq_for).CreateState();
+      for (int i = start; i < end; i++) {
+        int nq = num_qubits[i];
+        Simulator sim = Simulator(nq, tfq_for);
+        StateSpace ss = StateSpace(nq, tfq_for);
+        if (nq > largest_nq) {
+          // need to switch to larger statespace.
+          largest_nq = nq;
+          sv = ss.CreateState();
+        }
+        ss.SetStateZero(sv);
+        for (int j = 0; j < fused_circuits[i].size(); j++) {
+          qsim::ApplyFusedGate(sim, fused_circuits[i][j], sv);
+        }
+
+        for (uint64_t j = 0; j < (uint64_t(1) << nq); j++) {
+          (*output_tensor)(i, j) = ss.GetAmpl(sv, j);
+        }
+        for (uint64_t j = (uint64_t(1) << nq);
+             j < (uint64_t(1) << max_num_qubits); j++) {
+          (*output_tensor)(i, j) = std::complex<float>(-2, 0);
+        }
+      }
+      sv.release();
+    };
+
+    const int num_cycles = 200 * (1 << max_num_qubits);
```

Comment:
Wondering why this number is different than at https://github.com/tensorflow/quantum/blob/1bccab0e8751fa062b728ab150ac15420217a881/tensorflow_quantum/core/src/util_qsim.h#L56 ?

Reply:
For this simulation technique the unit of work is "the cost of simulating one circuit end to end and computing the expectation value" for the other the unit of work is "the time to update ONE amplitude entry associated with a certain index". One scales with system size while the other is a constant.